review comment changes

shikhin-metron opened the pull request:
Made the changes Jamie had suggested in the markdown file. Still getting the comments with multiple calls because when used the request with array of triage IDs, it gives variable comment ID. 
@ds-jdcockrill we can discuss further changes and comments here.

shikhin-metron committed:
review comment changes

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -18,22 +18,30 @@ def __init__(self, id, key, secret, url):
+        for ele in alert_ids[1:]:
+            alert_id_str = alert_id_str + "&id=" + ele
+
+        alert_url = self.url + "alerts?limit=20&id=" + str(alert_id_str)
```
if we have a specific list of IDs we don't need `limit` in the argument list.

we should also be able to assemble the url parameters as a dictionary and pass it in to `requests.get` as the `params` argument. See: https://docs.python-requests.org/en/latest/user/quickstart/#passing-parameters-in-urls

This should work fine with:

```
alert_url = self.url + "alerts"
params = dict(id=alert_ids)
response = requests.get(alert_url, params=params, headers=...)
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -18,22 +18,30 @@ def __init__(self, id, key, secret, url):
+            alert_id_str = alert_id_str + "&id=" + ele
+
+        alert_url = self.url + "alerts?limit=20&id=" + str(alert_id_str)
         response = requests.get(alert_url, headers={"Authorization": "Basic %s" % self.b64val, "searchlight-account-id": "%s" % self.id, "User-Agent": "DigitalShadowsAzureSentinelIntegration"})
```
you could remove a lot of duplicated code by using `requests.Session` to make your calls and put all these headers in there.

In the `__init__` method:

```
self.session = requests.Session()
self.session.headers.update({
  "Authorization": "Basic %s" % self.b64val,
  "searchlight-account-id": "%s" % self.id,
  "User-Agent": "DigitalShadowsAzureSentinelIntegration"
})
```

Then if you make your HTTP calls with `self.session.get(...)` (etc) then all those headers are added automatically.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -18,22 +18,30 @@ def __init__(self, id, key, secret, url):
+
+        alert_url = self.url + "alerts?limit=20&id=" + str(alert_id_str)
         response = requests.get(alert_url, headers={"Authorization": "Basic %s" % self.b64val, "searchlight-account-id": "%s" % self.id, "User-Agent": "DigitalShadowsAzureSentinelIntegration"})
         logging.info("Alerts response code: %s" % response.status_code)
```
We aren't testing the response code in any of our HTTP calls. If you aren't getting a 2XX status code response then we should probably be raising an error. The following can happen:

* if a triage-item or alert is deleted, you'll get a event of type 'update' but calling the endpoint to retrieve it (either the `/triage-item/{id}` or `/alert/{id}`) will return a 404 - probably not an issue here as you are using `/triage-items` and `/alerts` - but in those cases you likely won't receive an entry for the ID of deleted items
* if you breach the rate-limits on our API, you'll get a 429 response. In this case we either want to wait (there is a header called 'Retry-After' that says how long to wait for) - https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/429 - or just error out completely and start the loop again at the next poll

If we aren't getting a status code between 200 and 299 we should be erroring out and not continuing.

ds-jdcockrill replied:
@kapil-metron this one is important

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -18,22 +18,30 @@ def __init__(self, id, key, secret, url):
+        for ele in incident_ids[1:]:
+            inc_id_str = inc_id_str + "&id=" + str(ele)
+
+        incident_url = self.url + "incidents?limit=20&id=" + inc_id_str
```
as above - we don't need `limit` and can build the url-parameters using a dictionary

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -44,7 +52,7 @@ def get_triage_events(self, before_date, after_date):
         """
 
-        triage_url = self.url + "triage-item-events?event-created-before=" + str(before_date) + "&event-created-after=" +  str(after_date)
+        triage_url = self.url + "triage-item-events?limit=20&event-created-before=" + str(before_date) + "&event-created-after=" +  str(after_date)
```
can build the url-parameters using a dictionary

ds-jdcockrill replied:
resolving this as I've mentioned this in another comment

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -77,8 +85,7 @@ def get_triage_events_by_num(self, event):
         triage_url = self.url + "triage-item-events?limit=20&event-num-after=" + str(event)
         response = requests.get(triage_url, headers={"Authorization": "Basic %s" % self.b64val, "searchlight-account-id": "%s" % self.id})
-        logging.info("Events by num response code: %s" % response.status_code)
+        logging.info("Events by num %d response code: %s" % (event, response.status_code))
```
can we initialise a named logger in this module (and others as well) please? it'll make it easier to track down where logs are coming from.

For example:

```
import logging

logger = logging.getLogger('DS_api')

# within DS_api
  logging.info("my log message")
```

When we look at this in the logs, rather than getting something like:

```
root - my log message
```

we'll get

```
DS_api - my log message
```

So we'll know which class it came from

shikhin-metron replied:
This thing actually wont matter because in Sentinel it doesn't show it like `DS_api - my log message` neither on local run or on Azure portal. It still shows it like earlier somehow.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -43,26 +43,31 @@ def post_azure(self, response, item):
             posts to azure after appending triage information on it
         """
         json_obj = json.loads(response.text)
```
why are we doing `json.loads(response.text)` when requests provides `response.json()` ?

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -58,7 +66,7 @@ def get_triage_items(self, triage_ids):
+        items_url = self.url + "triage-items?limit=20&id=" + item_id_str
         response = requests.get(items_url, headers={"Authorization": "Basic %s" % self.b64val, "searchlight-account-id": "%s" % self.id})
         logging.info("Triage items response code: %s" % response.status_code)
         return response.text
```
there is a repeated pattern throughout `DS_api` and `DS_poller` where we are returning the response body as a string from `DS_api` and immediately pushing it through `json.loads()` in `DS_poller`.

It would be much neater if we were just returning the JSON object from `DS_api`. For instance, instead of:

```
return response.text
```

this:

```
return response.json()
```

That way we're not always calling `json.loads` around all our calls to `DS_api`.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -43,26 +43,31 @@ def post_azure(self, response, item):
-                comment_data_filtered.append(comment)
-
+        for i in range(len(json_obj)):
+            comment_data = json.loads(self.DS_obj.get_triage_comments(item[i]['id']))
```
i don't think it's guaranteed that the order of the elements in `item` are the same as the order of the elements in the response.

Indeed if something has been deleted (see comment above) then the response from `/alerts` or `/incidents` may contain fewer items. Thus we should go through both lists and match them up before merging their data together.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -73,29 +78,29 @@ def get_data(self):
                 event_data = json.loads(event_dataJSON)
-                self.date.post_event(self.event + len(event_data[:20]))
+                #calculating the max event number from current batch to  use in next call
+                max_event_num = max([e['event-num'] for e in event_data])
```
this will throw an error if `event_data` is an empty list - or it'll be `None`. You will need to check to see if any data has been returned before updating `max_event_num` - same applies in the `else` block

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -73,29 +78,29 @@ def get_data(self):
-        except (ValueError, IndexError, UnboundLocalError):
-            
+            logging.info(triage_id)
+        except (ValueError, IndexError, UnboundLocalError):            
```
we shouldn't need to catch these exceptions - if no new incidents or alerts have been found we can check that beforehand so we don't get an `IndexError`

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -73,29 +78,29 @@ def get_data(self):
+        except (ValueError, IndexError, UnboundLocalError):            
             logging.info("JSON is of invalid format or no new incidents or alerts are found")
         
         item_data = json.loads(self.DS_obj.get_triage_items(triage_id))
```
`triage_id` might be empty if we didn't get any data back from the API - so we shouldn't be calling `DS_obj.get_triage_items` with an empty list or we'll get something like an `IndexError`. If `triage_id` is empty then just `return [], None`

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -106,18 +111,25 @@ def poll(self):
-            item_data = self.get_data()
+            inc_ids = []
+            alert_ids = []
+            item_data, max_event_num = self.get_data()
```
if `item_data` is empty or `max_event_num is None` then we can just return without doing anything else since we have received no further data that needs processing.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -106,18 +111,25 @@ def poll(self):
+
+            #saving event num for next invocation
+            self.date.post_event(max_event_num)
         except (KeyError, TypeError, UnboundLocalError, IndexError):
```
we really shouldn't need to catch any of these exceptions - if any of these are thrown it is a programming error by us and we should fix it. For instance `KeyError` means we are trying to access a key in a dict before testing it exists. `IndexError` means we're trying to access an element in a list without testing if the list has enough (or any) elements in it.

I can imagine we could have a catch-all like this:

```
except Exception:
  logging.exception("Unexpected error occurred during processing")
```

`logging.exception` is useful because it prints the stack-trace to the logs which will help work out where the problem is.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -18,22 +18,30 @@ def __init__(self, id, key, secret, url):
         self.id = id
```
can we rename the folder this is in to get the casing of `Digital Shadows` right. Currently it is:

```
Data Connectors/Digital shadows/
```

and it should be (ideally)

```
Data Connectors/Digital Shadows/
```

shikhin-metron replied:
I have renamed the folder, but I think now if we create a PR it will be like a new Solution for the reviewers. This means they will check all of the things again and it would take lot of time like the current PR which is open for main Azure Sentinel repo.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -43,26 +43,31 @@ def post_azure(self, response, item):
-                comment_data_filtered.append(comment)
-
+        for i in range(len(json_obj)):
+            comment_data = json.loads(self.DS_obj.get_triage_comments(item[i]['id']))
```
Separate comment about retrieving triage item comments...

You can use this API endpoint to retrieve multiple comments in one call:

```
GET /triage-item-comments
```

and in the same way that you can get multiple triage items in one request with `/triage-items` with lots of `id` url parameters, the same is true of this endpoint. E.g.

```
/triage-item-comments?id=1&id=2&id=3
```

This would prevent you needing to make a separate call per triage item for comments.

shikhin-metron replied:
The request for getting multiple triage comments together has one issue where the comments IDs are not constant. If this does not matter to users, then I will get them all in one call as well.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -18,22 +18,30 @@ def __init__(self, id, key, secret, url):
         self.id = id
```
we haven't changed `AS_api` so I can't comment directly on the line, but on line 58 we have a print statement

```
print(response.text)
```

can we remove that please?

Also in the lines below that - if we don't get a success status code we should probably throw an exception. You can use `response.raise_for_status()` to do this automatically for you if you want.

This would avoid you from marking a batch as successful and incrementing the `event_num` - which is what will currently happen as we aren't throwing an exception if the request to Azure fails.

ds-jdcockrill replied:
in `Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsSearchlight_API_functionApp.json`

can we fix the capitalisation of `Digital shadows` please?

```
    "sampleQueries": [
        {
            "description" : "All Digital shadows incidents and alerts ordered by time most recently raised",
            "query": "DigitalShadows_CL | order by raised_t desc"
        }
    ],
```

and

```
			{
				"name": "REST API Credentials/permissions",
				"description": "**Digital shadows account ID, secret and key** is required.  See the documentation to learn more about API on the `https://portal-digitalshadows.com/learn/searchlight-api/overview/description`."
			}
```

Also in `readme.md` we have:

```
## Folders

1. `Digital shadows/` - This contains The package...
```

Also note that `The package` should be `the package`

ds-jdcockrill replied:
I've just attempted to deploy this manually into a new Azure Sentinel instance and I noticed that during the deployment the parameter names don't look great. I'm seeing things like

`Digitalshadows Account ID` rather than `Digital Shadows Account ID`

in `digitalshadowsARM.json` could we try changing these key names please?

```
        "digitalshadowsAccountID": {
            "type": "securestring",
            "defaultValue": ""
        },
        "digitalshadowsKey": {
            "type": "securestring",
            "defaultValue": ""
        },
        "digitalshadowsSecret": {
            "type": "securestring",
            "defaultValue": ""
        },
        "digitalshadowsUrl": {
            "type": "string",
            "defaultValue": ""
        },
        "historicaldays": {
```

And see if that fixes their naming when deploying? For instance, change: 

* `digitalshadowsAccountID` to `DigitalShadowsAccountID` 
* `digitalshadowsKey` to `DigitalShadowsKey`
* `digitalshadowsSecret` to `DigitalShadowsSecret`
* `digitalshadowsUrl` to `DigitalShadowsURL`
* `historicaldays` to `HistoricalDays`

this might necessitate changes to the data connector where it is importing those parameters from environment variables. I.e.

```
account_id = os.environ['digitalshadowsAccountID']
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -43,26 +43,31 @@ def post_azure(self, response, item):
             posts to azure after appending triage information on it
```
`from DigitalShadowsConnectorAzureFunction import state_serializer`

imported but not used - remove it if it isn't used

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -18,22 +18,30 @@ def __init__(self, id, key, secret, url):
         self.id = id
```
in the supplied requirements.txt can we remove urllib please? It is part of core python and can't be separately installed.

shikhin-metron replied:
Removing this from requirements causes an urllib3 missing error when deployed. We have seen this quite a lot when developing and testing. Not really sure why it happens.

ds-jdcockrill replied:
`urllib3` is different to `urllib` - you can put `urllib3` in a requirements file

ds-jdcockrill replied:
@shikhin-metron - could you try `urllib3` instead of `urllib` please and let me know if that works fine when deployed?

ds-jdcockrill reviewed (changes requested):
See other comments for requested changes.

shikhin-metron committed:
the requested changes

shikhin-metron committed:
added links pointing to this repo

shikhin-metron committed:
2 loops for json_obj and item + added handling exception if no new events are present

shikhin-metron commented:
@ds-jdcockrill this PR is ready for review. Please have a look at it and suggest any changes if required.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/AS_api.py:
```diff
@@ -55,11 +57,11 @@ def post_data(self, body, log_type):
+            logger.info('Accepted')
         else:
-            logging.info("Response code: {}".format(response.status_code))
+            response.raise_for_status()
```
I don't think we need an if-else here - we should just always call `response.raise_for_status()` as it only throws an error if one occurred. If we get a successful response then calling `raise_for_status()` doesn't do anything.

shikhin-metron replied:
Should we keep the if-block? We will know if the data got ingested in the logs when it prints 'Accepted'.

ds-jdcockrill replied:
if you want to log `Accepted`, do this:

```
response.raise_for_status()
log.info('Accepted')
```

You don't need the if block - if you look at the documentation and source for `raise_for_status` it will raise a `HttpError` if there was an error - so the `log.info('Accepted')` only gets logged if there was no error.

https://docs.python-requests.org/en/latest/_modules/requests/models/#Response.raise_for_status

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -0,0 +1,89 @@
+        self.session.headers.update({
+        "Authorization": "Basic %s" % self.b64val,
+        "searchlight-account-id": "%s" % self.id,
+        "User-Agent": "DigitalShadowsAzureSentinelIntegration"
```
fix indentation please - lines 24-26 should be indented 1 further step

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -0,0 +1,89 @@
+        params = dict(id=alert_ids)
+        response = self.session.get(alert_url, params=params)
+        logger.info("Alerts response code: %s" % response.status_code)
+        return response
```
all of these methods should be doing:

```suggestion
        response.raise_for_status()
        return response.json()
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -0,0 +1,89 @@
+        params = dict(id=incident_ids)
+        response = self.session.get(incident_url, params=params)
+        logger.info("Incident response code: %s" % response.status_code)
+        return response
```
```suggestion
        response.raise_for_status()
        return response.json()
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -0,0 +1,89 @@
+        triage_url = self.url + "triage-item-events?limit=20&event-created-before=" + str(before_date) + "&event-created-after=" +  str(after_date)
+        response = self.session.get(triage_url)
+        logger.info("Events response code: %s" % response.status_code)
+        return response.json()
```
```suggestion
        response.raise_for_status()
        return response.json()
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -0,0 +1,89 @@
+        params = dict(id=triage_ids)
+        response = self.session.get(items_url, params=params)
+        logger.info("Triage items response code: %s" % response.status_code)
+        return response.json()
```
```suggestion
        response.raise_for_status()
        return response.json()
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -0,0 +1,89 @@
+        items_url = self.url + "triage-items/" + str(item_id) + "/comments"
+        response = self.session.get(items_url)
+        logger.info("Comments response code: %s" % response.status_code)
+        return response.json()
```
```suggestion
        response.raise_for_status()
        return response.json()
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -0,0 +1,89 @@
+        triage_url = self.url + "triage-item-events?limit=20&event-num-after=" + str(event)
+        response = self.session.get(triage_url)
+        logger.info("Events by num %d response code: %s" % (event, response.status_code))
+        return response.json()
```
```suggestion
        response.raise_for_status()
        return response.json()
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+        """
+            posts to azure after appending triage information on it
+        """
+        json_obj = response.json()
```
if you make the suggested changes above, then this `response.json()` is not needed.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+        return res
+
+
+    def post_azure(self, response, item):
```
these argument names could be better, e.g. instead of:

`response, item`

how about:

`alerts_and_incidents, triage_items`

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+        """
+        json_obj = response.json()
+        for it in item:
+            for i in range(len(json_obj)):
```
`enumerate` is designed to make this easier:

```py
            for i, obj in enumerate(json_obj):
```
but to be honest what you are using it for doesn't look correct anyway. I think I said previously that we needed to ensure that we validate that the IDs match in the two lists and I think this is probably better achieved by grouping them together in advance and then looping through that. For instance:

```py
data = {}
for item in triage_items:
  # in this loop i'm storing a tuple of 2 elements, one for the triage item
  # data and another (empty) entry for the alert/incident data that it found
  # in the next loop
  if 'alert-id' in item['source']:
    data[item['source']['alert-id']] = (item,None)
  elif 'incident-id' in item['source']:
    data[item['source']['incident-id']] = (item,None)
  else:
    raise Exception(f'Triage item missing expected source ID field: {item}')

for item in alerts_and_incidents:
  # find the dict entry with the matching key and add ourselves to the
  # second element in the tuple
  if item['id'] in data:
    data[item['id']][1] = item
  else:
    raise Exception(f'No matching triage item found for alert/incident: {item}')

# now the loop as you had it
for triage_item, alert_or_incident in data.values():
  # validate we actually have an alert/incident before proceeding
  if not alert_or_incident:
    raise Exception(f'Triage item had no matching alert/incident data: {triage_item}')
  alert_or_incident['status'] = triage_item['state']
  alert_or_incident['triage_id'] = triage_item['id']
  # etc for your other code```

shikhin-metron replied:
Changing `data[item['id']][1] = item` line, will have to convert tuple to list and back to tuple or it wont let me run that line, since tuples are immutable. It should work like this, right?

ds-jdcockrill replied:
You could leave it as a list - I wrote the code without running it so it's entirely possible that I missed the fact that tuples are immutable.

You could also replace the tuple with:

```
for item in alerts_and_incidents:
  # find the dict entry with the matching key and add ourselves to the
  # second element in the tuple
  if item['id'] in data:
    data[item['id']] = (data[item['id']][0], item)
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+            for i in range(len(json_obj)):
+                if(it['source']['alert-id'] == json_obj[i]['id'] or it['source']['incident-id'] == json_obj[i]['id']):
+
+                    comment_data = self.DS_obj.get_triage_comments(it['id'])
```
i'd move this down to just above where you start looping around the comment data

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+                    json_obj[i]['status'] = it['state']
+                    json_obj[i]['triage_id'] = it['id']
+                    json_obj[i]['triage_raised_time'] = it['raised']
+                    json_obj[i]['triage_updated_time'] = it['updated']
```
i might've done this flow differently - rather than adding data onto the `alert_or_incidents` object I would've created a new object with data copied onto it:

```py
azure_obj = {
  **alert_or_incident,
  'status': triage_item['state'],
  'triage_id': triage_item['id'] # etc for others
}
```

This avoids mutating the object passed into the function, which could have side-effects if the caller tries to use that object for something else later.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+                        del comment['updated']
+                        del comment['user']
+                        if comment['content'] != "":
+                            comment_data_filtered.append(comment)
```
Similar comment to the above about copying rather than mutating. Also if you use my structure above:

```py
azure_obj = {
  ...alert_or_incident,
  status: triage_item['state'],
  triage_id: triage_item['id'],
  comments: [] # empty list here
}
```

then this part becomes:

```py
                    for comment in comment_data:
                        if not comment['content']: # may as well do this first
                          continue
                        azure_obj['comments'].append({
                          'user_name': comment['user']['name'],
                          'content': comment['content'],
                          'id': comment['id'],
                          'created': comment['created']
                        })
```

Another benefit of doing this is that if we add more field they aren't automatically dragged into the integration - which gives us a bit of control as to how they are introduced.

I also noticed that we are being inconsistent over our formatting - in the existing code we have `user-name` as a field whereas elsewhere we've used underscores, e.g. `triage_item_id`. Ideally we align them.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+            if(isinstance(self.event, int)):
+                event_data = self.DS_obj.get_triage_events_by_num(self.event)
+                #calculating the max event number from current batch to  use in next call
+                if(event_data):
```
don't need brackets round this:

```suggestion
                if event_data:
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+            inc_ids = []
+            alert_ids = []
+            item_data, max_event_num = self.get_data()
+            if(item_data):
```
don't need brackets here;

```suggestion
            if item_data:
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+            else:
+                event_data = self.DS_obj.get_triage_events(self.before_time, self.after_time)
+                #calculating the max event number from current batch to  use in next call
+                if(event_data):
```
```suggestion
                if event_data:
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+                    logger.info("Total number of events are " + str(len(event_data)))
+                else:
+                    logger.info("There are no new events")
+                    return [], int
```
shouldn't return the int object here.

Perhaps initialise `max_event_num` to `-1` at the start of the method and if then we don't need this else statement.

In the calling method we can look at `max_event_num` and if it is `-1` we know there's nothing to look at. Although I can see you are already checking to see if there are any items in `event_data`

We probably just remove this `else` block - you don't have a matching on in the `if` block above and don't need it here..

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+                    triage_id.append(event['triage-item-id'])
+
+            logger.info(triage_id)
+        except Exception:            
```
I don't think we should be catching this at all. We should allow exceptions to propagate up to `poll`. As it stands we'll continue and attempt to `get_triage_items` with an empty list of triage IDs if we catch here.

ds-jdcockrill replied:
@shikhin-metron - this `except` block still seems to be here - if we get an exception during `get_triage_events_by_num` then we shouldn't catch it here - we should just let it propagate up to the `poll` method which will log it there and exti.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,155 @@
+
+                #saving event num for next invocation
+                self.date.post_event(max_event_num)
+        except Exception:
+            if(item_data):
+                logger.exception("Error polling: ")
+            else:
+                logger.info("No new events found.")
```
this log statement belongs in an `else` block alongside `if item_data`, not in a catch block.

ds-jdcockrill replied:
```suggestion
        except Exception:
            logger.exception("Error polling")
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/digitalshadowsARM.json:
```diff
@@ -16,23 +16,23 @@
             "defaultValue": ""
         },
-        "digitalshadowsUrl": {
+        "DigitalShadowsUrl": {
```
```suggestion
        "DigitalShadowsURL": {
```

shikhin-metron committed:
made the requested changes

shikhin-metron commented:
Need some more time, a bug came up.

shikhin-metron committed:
fixed some issues which were not caught before last commit

shikhin-metron commented:
@ds-jdcockrill PR is ready for review.

Added limit to listing triage items because default was 10 and we were losing data. Fixed some other issues which I did not catch before while testing.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -0,0 +1,90 @@
+            send only the DS converted dates using state serializer functions to get triage events
+        """
+
+        triage_url = self.url + "triage-item-events?limit=20&event-created-before=" + str(before_date) + "&event-created-after=" +  str(after_date)
```
For consistency, can we change this to using a `params` dict like the other methods please?

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_api.py:
```diff
@@ -0,0 +1,90 @@
+        """
+            gets triage events by number
+        """
+        triage_url = self.url + "triage-item-events?limit=20&event-num-after=" + str(event)
```
For consistency, can we change this to using a `params` dict like the other methods please?

shikhin-metron replied:
params for the limit?

ds-jdcockrill replied:
and the `event-num-after` stuff, i.e.

```py
params = dict(limit=20)
params['event-num-after']=str(event)
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,173 @@
+        event_data = []
+
+        try:
+            if(isinstance(self.event, int)):
```
```suggestion
            if isinstance(self.event, int):
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,173 @@
+                    logger.info("Total number of events are " + str(len(event_data)))
+            
+            for event in event_data:
+                if(event is not None and event['triage-item-id'] not in triage_id):
```
```suggestion
                if event is not None and event['triage-item-id'] not in triage_id:
```

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,173 @@
+        except Exception:            
+            logger.exception("Error while getting triage data: ")
+        
+        if event_data:
```
shouldn't this be `if triage_id:` ?

We aren't using `event_data` in the body of the if-block.

shikhin-metron replied:
It is checking if we have any events, if there are no events then no need to get triage items. I can use triage_id instead as well I guess since it seems more intuitive. Will change it.

ds-jdcockrill commented on Solutions/Digital Shadows/Data Connectors/Digital Shadows/DigitalShadowsConnectorAzureFunction/DS_poller.py:
```diff
@@ -0,0 +1,173 @@
+        self.date = State(connection_string)
+        logger.info("got inside the poller code")
+        self.event = self.date.get_last_event(historical_days)
+        if(isinstance(self.event, tuple)):
```
```suggestion
        if isinstance(self.event, tuple):
```

shikhin-metron committed:
some more code review changes

shikhin-metron committed:
missed a small change

shikhin-metron committed:
added urllib3 in requirements, made requested changes and created package
